CRM/Contribute/Page/Tab.php

```diff
@@ -94,6 +94,87 @@ public static function recurLinks(int $recurID, $context = 'contribution') {
     return $links;
   }
 
+  /**
+   * Get the recur links to return for self service.
+   *
+   * These are the links to present to a logged in user wishing
+   * to service their own
+   *
+   * @param int $recurID
+   *
+   * @return array|array[]
+   * @throws \CRM_Core_Exception
+   * @throws \CiviCRM_API3_Exception
+   */
+  public static function selfServiceRecurLinks(int $recurID): array {
+    $links = [];
+    $paymentProcessorObj = Civi\Payment\System::singleton()->getById(CRM_Contribute_BAO_ContributionRecur::getPaymentProcessorID($recurID));
+    if ($paymentProcessorObj->supports('cancelRecurring')
+      && $paymentProcessorObj->subscriptionURL($recurID, 'recur', 'cancel')
+    ) {
+      $url = $paymentProcessorObj->subscriptionURL($recurID, 'recur', 'cancel');
+      $links[CRM_Core_Action::DISABLE] = [
+        'url' => $url,
+        'name' => ts('Cancel'),
+        'title' => ts('Cancel'),
+        // Only display on-site links in a popup.
+        'class' => (stripos($url, 'http') !== FALSE) ? 'no-popup' : '',
```

Comment:
Hmm, looks like on-site links (eg dummy processor) get this class as well due to the absolute URL returned by `subscriptionURL()` in the base file - https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/Payment.php#L1768.